### DISCUSSION TITLE
Newbie question: How do i persist hyprland config changes between rebuilds?

### DISCUSSION THREAD

**Discussion by Razkaroth** [richen604/hydenix#7] [Q&A]
Hi!
I just moved to hydenix and it looks and feels great.
Thing is, whenever I rebuild the system, all hyprland changes I made are lost (monitors, extra keybinds, etc.).
Is there a way to persist them, or to declare them on the configuration?

I'm thinking of writing something that overwrites the contents at the end of the system build, but i am not sure where in the config should i make those changes so that they are not overwrited.

**Comment by richen604** [comment]
@Razkaroth missed you asked this here as well as on the discord. let me know if you need any more help with your current configuration

> I'm thinking of writing something that overwrites the contents at the end of the system build, but i am not sure where in the config should i make those changes so that they are not overwrited.

this is my goal with the release currently in dev branch. having good defaults for users to override with a flake template. I also am building my personal configuration to be better used as an example. stay tuned for that

**Comment by AdrienLemaire** [comment]
I would love to understand how to persist config changes as well, if you don't mind sharing here as well.

**Reply by richen604** [reply]
> I would love to understand how to persist config changes as well, if you don't mind sharing here as well.

what exactly are you trying to persist? 
in home-manager you mostly use `home.file`. You can see the [template readme](https://github.com/richen604/hydenix/blob/main/template/README.md) for some common examples

**Reply by AdrienLemaire** [reply]
Thanks @richen604 , I had missed this readme, as it is only referenced in the main readme's Contributing section.

In [home.file](https://github.com/richen604/hydenix/tree/main/template#homefile), I couldn't figure out the `Then in your config.nix homeModules import` part.

```nix
# config.nix
...
  homeModules = [
    #(toString ./home-module.nix)
  ];
  home.file = {
    ".config/hypr/keybindings.conf" = {
        source = ./hypr-keybindings.conf;
        force = true;
        mutable = true;
    };
  };
```
I tried to add a home-module.nix file with the same code inside, but that threw me some error
```
       error: syntax error, unexpected '=', expecting end of file
       at /nix/store/dkjzrzcgl05g5h74wwavsvf85bljnb4d-source/home-module.nix:1:11:
            1| home.file = {
             |           ^
            2|   ".config/hypr/keybindings.conf" = {
```

**Reply by richen604** [reply]
both homeModules and nixosModules are a list of modules, essentially lamdas. 
what nix is complaining about is a syntax issue. how you would write this is

```nix
  homeModules = [
     (
      { userConfig, pkgs, ... }:
      {
        home.file = {
          ".config/hypr/keybindings.conf" = {
            source = ./hypr-keybindings.conf;
            force = true;
            mutable = true;
          };
        };
      }
    )
  ];
  ```
  
 the above lamda could be in a file `home.nix` as well for simplicity

**Reply by AdrienLemaire** [reply]
Awesome. 
Thanks for taking the time :heart: 
I'll go and study nix in depth so that I can figure this stuff out going forwards :sweat_smile:

**Reply by AdrienLemaire** [reply]
I'm facing another issue.
Can you explain me how I can persist software that doesn't have an official package, but has its own flake configuration ?

I'd like to set ghostty as default terminal, and [their doc](https://ghostty.org/docs/install/binary#nix-flake) explains how to update a flake.

I went and modified the inputs and outputs of the `flake.nix` file from the hydenix repo, and rebuild works, but the `ghostty` executable is not available after rebuild.

<details><summary>flake.nix</summary>

```nix
{
  description = "template for hydenix";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    hydenix = {
      # Available inputs:
      # Main: github:richen604/hydenix
      # Dev: github:richen604/hydenix/dev 
      # Commit: github:richen604/hydenix/<commit-hash>
      # Version: github:richen604/hydenix/v1.0.0
      url = "github:richen604/hydenix";
    };

    ghostty = {
      url = "github:ghostty-org/ghostty";
    };
  };

  outputs =
    { nixpkgs, hydenix, ghostty, ...  }@inputs:
    let
      system = "x86_64-linux";

      hydenixConfig = hydenix.lib.mkConfig {
        userConfig = import ./config.nix;
        # inputs without nixpkgs to prevent override
        extraInputs = removeAttrs inputs [ "nixpkgs" ];
      };
    in
    {
      nixosConfigurations.${hydenixConfig.userConfig.host} = hydenixConfig.nixosConfiguration // {
        environment.systemPackages = with (import inputs.nixpkgs { system = system; }); [
          ghostty.packages.${system}.default
        ];
      };

      packages.${system} = {
        # Packages below load your config in ./config.nix

        # defaults to nix-vm - nix run .
        default = hydenixConfig.nix-vm.config.system.build.vm;

        # NixOS build packages - nix run .#hydenix / sudo nixos-rebuild switch/test --flake .#hydenix
        hydenix = hydenixConfig.nixosConfiguration.config.system.build.toplevel;

        # Home activation packages - nix run .#hm / nix run .#hm-generic / home-manager switch/test --flake .#hm or .#hm-generic
        hm = hydenixConfig.homeConfigurations.${hydenixConfig.username}.activationPackage;
        hm-generic = hydenixConfig.homeConfigurations."${hydenixConfig.username}-generic".activationPackage;

        # EXPERIMENTAL VM BUILDERS - nix run .#arch-vm / nix run .#fedora-vm
        arch-vm = hydenixConfig.arch-vm;
        #fedora-vm = hydenixConfig.fedora-vm;
      };

    };
}
```

</details>

**Reply by richen604** [reply]
you declare ghostyy as inputs in flake, since mkConfig is a function that takes inputs you can reference these in your configurations
in your `config.nix` add a module in `nixosModules` called `configuration.nix`

this most likely will not make ghostyy the default terminal, set ghostyy in as `TERM` env var
```nix
{
  inputs,
  config,
  userConfig,
  pkgs,
  ...
}:
{
  environment.systemPackages = [
    inputs.ghostty.packages.${system}.default
    ];
}
```
![image](https://github.com/user-attachments/assets/b0a04ed4-d60f-4c41-a234-67d6f623cee1)

**Reply by AdrienLemaire** [reply]
By setting the default term, I meant the hyperland binding `$mainMod+T` :smile: 

Lovely, works well now, thanks again !

**Reply by saharshbhansali** [reply]
> both homeModules and nixosModules are a list of modules, essentially lamdas. what nix is complaining about is a syntax issue. how you would write this is
> 
> ```nix
>   homeModules = [
>      (
>       { userConfig, pkgs, ... }:
>       {
>         home.file = {
>           ".config/hypr/keybindings.conf" = {
>             source = ./hypr-keybindings.conf;
>             force = true;
>             mutable = true;
>           };
>         };
>       }
>     )
>   ];
> ```
> 
> the above lamda could be in a file `home.nix` as well for simplicity

I've done this for a whole list of files, and a `nixos-rebuild` still complains about random files like `~/.config/Kvantum` and other files, that I don't even set in my custom config persistence file.
```
(ref for Kvantum error)
error: builder for '/nix/store/grwrpqvf94laps0fsx5safvhrn5fgas4-home-manager-files.drv' failed with exit code 1;
       last 6 log lines:
       > File conflict for file '.config/Kvantum/wallbash/wallbash.kvconfig'
       > File conflict for file '.config/Kvantum/wallbash/wallbash.svg'
       > File conflict for file '.config/hyde/hyde.conf'
       > File conflict for file '.config/hypr/themes/theme.conf'
       > File conflict for file '.config/hypr/userprefs.conf'
       > Error installing file '.config/kitty/kitty.conf' outside $HOME
       For full logs, run 'nix log /nix/store/grwrpqvf94laps0fsx5safvhrn5fgas4-home-manager-files.drv'.
error: 1 dependencies of derivation '/nix/store/i0l130w3kb7lgw3ng0vc1dbb2mzp03ms-home-manager-generation.drv' failed to build
```

I would like a few files to refer to my customized versions (like `~/.config/hypr/{userprefs,hyprlock,hypridle}.conf` and the directory `~/.config/kitty`) and others to work as intended in hydenix.

I've been forced to set things up as such, cause nixos complains if I don't use `lib.mkForce` (which I suspect is because the flake itself uses mkForce to set the configs in the first place).
`flake.nix`:
```nix
home-modules = [
  (toString ./home.nix)
];
```
and `home.nix`
```
  home.file =
    let
      confDir = "${config.home.homeDirectory}/.config";
      homeDir = "${config.home.homeDirectory}";
    in 
    {

      ".config/hypr/userprefs.conf" = lib.mkForce {
        source = <path>;
        force = true;
        mutable = true;
      };
      
      ".config/kitty" = lib.mkForce {
        source = "${confDir}/kitty"; 
        force = true;
        mutable = true;
        recursive = true;
      };

    };
```
where `path = "${confDir}/hypr/userprefs.conf"`

PS: I also have to use `--impure` while evaluating the files because I get an error stating that I cannot refer to `/home/` directly in pure evaluation mode.

**Reply by richen604** [reply]
you cant source files in your home directory (out of nix store) as a source for files in a pure way. `mkOutOfStoreSymlink` exists but i would advise against this

add them to your flake folder and use relative paths
this might also solve your kvantum error as I've never tested impure builds 

home.file calls in nix dont use mkForce. its that youre overriding this (i do this too)
